### PR TITLE
Modularize `:configuration-cache`

### PR DESCRIPTION
**Pull request by bamboo**
Initially by extracting three modules:
- `:configuration-problems-base`: base utilities for reporting configuration problems with structured messages and property traces
- `:object-serialization`: core configuration cache serialization api and implementation but no codecs
- `:java-language-extensions-kt`: Java/Kotlin stdlib extensions shared by all modules

This is just the first step, subsequent PRs will detangle and move more code from `:configuration-cache` into `:object-serialization` and rename all moved packages to match the new structure.

### Context

`:configuration-cache` is currently one of the largest Gradle modules. It is time to modularize it to reduce compilation times and cognitive overhead.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test ACC

File: platforms/core-configuration/kotlin-language-extensions/build.gradle.kts [hunk 1]
```diff
@@ -14,11 +14,13 @@
...
+    id("gradlebuild.distribution.implementation-kotlin")
+}
 
+description = "Useful extension methods for Kotlin"
 
-interface BeanStateWriter {
-    suspend fun WriteContext.writeStateOf(bean: Any)
+dependencies {
+    api(libs.futureKotlin("stdlib"))
+    implementation(projects.javaLanguageExtensions)
```
Comment (abstratt) [inline platforms/core-configuration/kotlin-language-extensions/build.gradle.kts:25]: Now that I see the parallel between this new module and the existing `java-language-extensions`, I have no issue with the new name. Actually, I think this module is a much better described as "stdlib-extensions" than the one for Java, as the latter actually introduces things that go way beyond providing general purpose extensions to the Java library. 

It seems we are consuming `java-language-extensions` just to get access to `Cast`. That is fine, more of a problem with that module's purpose than with this one (and that module's scope is already an improvement over recent past).

File: platforms/core-configuration/configuration-cache/build.gradle.kts [hunk 2]
```diff
@@ -40,8 +40,8 @@ tasks.withType<org.jetbrains.kotlin.gradle.tasks.KotlinCompile>().configureEach
 dependencies {
     api(projects.concurrent)
     api(projects.javaLanguageExtensions)
-    api(projects.serialization)
     api(projects.serviceProvider)
+    api(projects.configurationProblemsBase)
     api(project(":base-services"))
     api(project(":build-operations"))
     // TODO - it might be good to allow projects to contribute state to save and restore, rather than have this project know about everything
```
Comment (abstratt) [inline platforms/core-configuration/configuration-cache/build.gradle.kts:47]: Ah, this refers to the design you mentioned we want to pursue, which would de-invert the dependencies.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Contexts.kt [hunk 3]
```diff
@@ -98,56 +89,10 @@ class DefaultWriteContext(
 }
 
 
-internal
-class LoggingTracer(
```
Comment (abstratt) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Contexts.kt:102]: "deleted" classes here were moved to their own files

File: platforms/core-configuration/object-graph-serialization/src/main/kotlin/org/gradle/configurationcache/serialization/Combinators.kt [hunk 4]
```diff
@@ -18,71 +18,24 @@ package org.gradle.configurationcache.serialization
...
 import kotlin.coroutines.Continuation
 import kotlin.coroutines.CoroutineContext
 import kotlin.coroutines.coroutineContext
 import kotlin.coroutines.startCoroutine
 import kotlin.coroutines.suspendCoroutine
 
 
-internal
 fun <T> singleton(value: T): Codec<T> =
     SingletonCodec(value)
```
Comment (abstratt) [inline platforms/core-configuration/object-graph-serialization/src/main/kotlin/org/gradle/configurationcache/serialization/Combinators.kt]: functions were moved to platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/Unsupported.kt

File: platforms/core-configuration/object-graph-serialization/src/main/kotlin/org/gradle/configurationcache/serialization/Combinators.kt [hunk 5]
```diff
@@ -275,104 +219,34 @@ suspend fun <K, V, T : MutableMap<K, V>> ReadContext.readMapEntriesInto(items: T
 }
 
 
-internal
```
Comment (abstratt) [inline platforms/core-configuration/object-graph-serialization/src/main/kotlin/org/gradle/configurationcache/serialization/Combinators.kt]: Extension functions moved to platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/ClassPathEncodingExtensions.kt

File: platforms/core-configuration/object-graph-serialization/src/main/kotlin/org/gradle/configurationcache/serialization/Codec.kt [hunk 6]
```diff
@@ -110,39 +111,12 @@ interface IsolateContext {
 }
 
 
-sealed class IsolateOwner {
```
Comment (abstratt) [inline platforms/core-configuration/object-graph-serialization/src/main/kotlin/org/gradle/configurationcache/serialization/Codec.kt]: Moved to own file and made part of a new sealed class hierachy

**Review by abstratt — APPROVED** [review]
Good step in the right direction.

Only actual changes I could see were the introcdution of a base `ProblemsListener` implementation (`AbstractProblemsListener`), and introduction of a sealed class hierarchy for `IsolateOwner` implementations.

**Pull request merged**